From cb8e4ce2201daecf2e0a8ef046d4b7dda8a94b68 Mon Sep 17 00:00:00 2001
From: Masahiro Nakamura <tsuucat@icloud.com>
Date: Thu, 5 Aug 2021 20:55:41 +0900
Subject: [PATCH] no-titlebar-head

---
 src/nsterm.m | 9 +--------
 1 file changed, 1 insertion(+), 8 deletions(-)

diff --git a/src/nsterm.m b/src/nsterm.m
index 3676418c9b..71fd71522d 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -433,10 +433,7 @@ - (NSColor *)colorUsingDefaultColorSpace
 
 /* These flags will be OR'd or XOR'd with the NSWindow's styleMask
    property depending on what we're doing.  */
-#define FRAME_DECORATED_FLAGS (NSWindowStyleMaskTitled              \
-                               | NSWindowStyleMaskResizable         \
-                               | NSWindowStyleMaskMiniaturizable    \
-                               | NSWindowStyleMaskClosable)
+#define FRAME_DECORATED_FLAGS NSWindowStyleMaskBorderless
 #define FRAME_UNDECORATED_FLAGS NSWindowStyleMaskBorderless
 
 /* TODO: Get rid of need for these forward declarations.  */
@@ -8294,10 +8291,6 @@ - (instancetype) initWithEmacsFrame:(struct frame *)f
       if ([col alphaComponent] != (EmacsCGFloat) 1.0)
         [self setOpaque:NO];
 
-      /* toolbar support */
-      if (! FRAME_UNDECORATED (f))
-        [self createToolbar:f];
-
       /* macOS Sierra automatically enables tabbed windows.  We can't
          allow this to be enabled until it's available on a Free system.
          Currently it only happens by accident and is buggy anyway.  */
-- 
2.28.0

