From 58edc380df99983ad670338148196bc3241c636d Mon Sep 17 00:00:00 2001
From: Boris Buliga <boris@d12frosted.io>
Date: Wed, 5 Oct 2022 14:22:00 +0300
Subject: [PATCH] remove title bar but keep corners rounded

---
 src/nsterm.m | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/nsterm.m b/src/nsterm.m
index 1643ba5427..ec57bc0b7d 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -434,6 +434,7 @@ - (NSColor *)colorUsingDefaultColorSpace
 /* These flags will be OR'd or XOR'd with the NSWindow's styleMask
    property depending on what we're doing.  */
 #define FRAME_DECORATED_FLAGS (NSWindowStyleMaskTitled              \
+                               | NSFullSizeContentViewWindowMask    \
                                | NSWindowStyleMaskResizable         \
                                | NSWindowStyleMaskMiniaturizable    \
                                | NSWindowStyleMaskClosable)
@@ -8323,12 +8324,19 @@ - (instancetype) initWithEmacsFrame:(struct frame *)f
 #endif
     }
 
+ [self setTitlebarAppearsTransparent:YES];
+ [self setTitleVisibility:NSWindowTitleHidden];
+ [[self standardWindowButton:NSWindowCloseButton] setHidden:YES];
+ [[self standardWindowButton:NSWindowMiniaturizeButton] setHidden:YES];
+ [[self standardWindowButton:NSWindowZoomButton] setHidden:YES];
+
   return self;
 }
 
 
 - (void)createToolbar: (struct frame *)f
 {
+  return;
   if (FRAME_UNDECORATED (f) || !FRAME_EXTERNAL_TOOL_BAR (f))
     return;
 
-- 
2.36.1

