diff --git a/src/nsterm.m b/src/nsterm.m
index 3676418c9b..9599ee933e 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -433,9 +433,9 @@ - (NSColor *)colorUsingDefaultColorSpace
 
 /* These flags will be OR'd or XOR'd with the NSWindow's styleMask
    property depending on what we're doing.  */
-#define FRAME_DECORATED_FLAGS (NSWindowStyleMaskTitled              \
-                               | NSWindowStyleMaskResizable         \
-                               | NSWindowStyleMaskMiniaturizable    \
+#define FRAME_DECORATED_FLAGS (NSWindowStyleMaskBorderless       \
+                               | NSWindowStyleMaskResizable      \
+                               | NSWindowStyleMaskMiniaturizable \
                                | NSWindowStyleMaskClosable)
 #define FRAME_UNDECORATED_FLAGS NSWindowStyleMaskBorderless
 
@@ -8294,10 +8294,6 @@ - (instancetype) initWithEmacsFrame:(struct frame *)f
       if ([col alphaComponent] != (EmacsCGFloat) 1.0)
         [self setOpaque:NO];
 
-      /* toolbar support */
-      if (! FRAME_UNDECORATED (f))
-        [self createToolbar:f];
-
       /* macOS Sierra automatically enables tabbed windows.  We can't
          allow this to be enabled until it's available on a Free system.
          Currently it only happens by accident and is buggy anyway.  */
