From 04593253983e0c10f67cc7e9b0052f72f8fc1c7c Mon Sep 17 00:00:00 2001
From: Golem <golem@d12frosted.io>
Date: Thu, 9 Jan 2020 07:22:17 +0200
Subject: [PATCH] [patch] fix-window-role

---
 src/nsterm.m | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/nsterm.m b/src/nsterm.m
index a029af5a69..d57b558350 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -8755,7 +8755,7 @@ - (id)accessibilityAttributeValue:(NSString *)attribute
   NSTRACE ("[EmacsWindow accessibilityAttributeValue:]");
 
   if ([attribute isEqualToString:NSAccessibilityRoleAttribute])
-    return NSAccessibilityTextFieldRole;
+    return NSAccessibilityWindowRole;
 
   if ([attribute isEqualToString:NSAccessibilitySelectedTextAttribute]
       && curbuf && ! NILP (BVAR (curbuf, mark_active)))
-- 
2.26.0

