From 9f4e7f80c8ed165e8a15fc91c95b3777df766893 Mon Sep 17 00:00:00 2001
From: Golem <golem@d12frosted.io>
Date: Thu, 9 Jan 2020 07:21:55 +0200
Subject: [PATCH] [patch] no-frame-refocus-cocoa

---
 src/frame.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/src/frame.c b/src/frame.c
index c871e4fd99..470ab948bd 100644
--- a/src/frame.c
+++ b/src/frame.c
@@ -2056,15 +2056,6 @@ delete_frame (Lisp_Object frame, Lisp_Object force)
 		}
 	    }
 	}
-#ifdef NS_IMPL_COCOA
-      else
-	/* Under NS, there is no system mechanism for choosing a new
-	   window to get focus -- it is left to application code.
-	   So the portion of THIS application interfacing with NS
-	   needs to know about it.  We call Fraise_frame, but the
-	   purpose is really to transfer focus.  */
-	Fraise_frame (frame1);
-#endif
 
       do_switch_frame (frame1, 0, 1, Qnil);
       sf = SELECTED_FRAME ();
-- 
2.26.0

