From eca8d174eefe8e2c0b09a2488802b90cfb002ed9 Mon Sep 17 00:00:00 2001
From: Boris Buliga <boris@d12frosted.io>
Date: Sat, 4 Jan 2020 15:55:21 +0200
Subject: [PATCH] [patch] no-frame-refocus-cocoa

---
 src/frame.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/src/frame.c b/src/frame.c
index 88d6f22fc0..4ab5787b8d 100644
--- a/src/frame.c
+++ b/src/frame.c
@@ -2056,15 +2056,6 @@ delete_frame (Lisp_Object frame, Lisp_Object force)
 		}
 	    }
 	}
-#ifdef NS_IMPL_COCOA
-      else
-	/* Under NS, there is no system mechanism for choosing a new
-	   window to get focus -- it is left to application code.
-	   So the portion of THIS application interfacing with NS
-	   needs to know about it.  We call Fraise_frame, but the
-	   purpose is really to transfer focus.  */
-	Fraise_frame (frame1);
-#endif
 
       do_switch_frame (frame1, 0, 1, Qnil);
       sf = SELECTED_FRAME ();
-- 
2.24.1

