From 11f74c32af3f2f5385407672cdeabcd2b9079958 Mon Sep 17 00:00:00 2001
From: Boris Buliga <boris@d12frosted.io>
Date: Thu, 24 May 2018 21:19:03 +0300
Subject: [PATCH] increase size of remembered_data

---
 src/pdumper.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/pdumper.c b/src/pdumper.c
index 400177223a..ebcacf5a7d 100644
--- a/src/pdumper.c
+++ b/src/pdumper.c
@@ -136,5 +136,5 @@ static struct
 {
   void *mem;
   int sz;
-} remembered_data[32];
+} remembered_data[64];
 static int nr_remembered_data = 0;

 typedef int32_t dump_off;
--
2.17.0
