From 7b46b8e7993b0ecd542cae27162bd3ed589e6e26 Mon Sep 17 00:00:00 2001
From: Golem <golem@d12frosted.io>
Date: Wed, 31 Mar 2021 09:03:34 +0300
Subject: [PATCH] [PATCH] fix code signing on arm macos systems

---
 src/Makefile.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Makefile.in b/src/Makefile.in
index 39fd708564..c66e67447a 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -339,7 +339,7 @@ HAVE_PDUMPER = @HAVE_PDUMPER@
 
 ## ARM Macs require that all code have a valid signature.  Since pump
 ## invalidates the signature, we must re-sign to fix it.
-DO_CODESIGN=$(patsubst arm-apple-darwin%,yes,@configuration@)
+DO_CODESIGN=$(patsubst aarch64-apple-darwin%,yes,@configuration@)
 
 # 'make' verbosity.
 AM_DEFAULT_VERBOSITY = @AM_DEFAULT_VERBOSITY@
-- 
2.30.0

