From d3c9dcd4cb1d30a27e1e618ea5a9f5c12e103720 Mon Sep 17 00:00:00 2001
From: Boris Buliga <boris@d12frosted.io>
Date: Tue, 9 Aug 2022 09:17:27 +0300
Subject: [PATCH] fix MAC_LIBS inference after gcc 12 release

---
 configure.ac | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/configure.ac b/configure.ac
index 1a264275bd..40777ba002 100644
--- a/configure.ac
+++ b/configure.ac
@@ -4240,8 +4240,7 @@ if test "${with_native_compilation}" != "no"; then
         if test -n "`$BREW --prefix --installed libgccjit 2>/dev/null`"; then
           MAC_CFLAGS="-I$(dirname $($BREW ls -v libgccjit | \
                                                 grep libgccjit.h))"
-          MAC_LIBS="-L$(dirname $($BREW ls -v libgccjit| \
-                                            grep -E 'libgccjit\.(so|dylib)$'))"
+          MAC_LIBS="-L$(dirname $($BREW ls -v libgccjit | grep -E 'libgccjit\.(so|dylib)$' | tail -1))"
         fi
       fi
 
-- 
2.36.1

