diff --git a/nextstep/templates/Info.plist.in b/nextstep/templates/Info.plist.in
index beddbf4a893..38b8c083dbb 100644
--- a/nextstep/templates/Info.plist.in
+++ b/nextstep/templates/Info.plist.in
@@ -693,5 +693,7 @@ along with GNU Emacs.  If not, see <https://www.gnu.org/licenses/>.
         <string>Emacs requires permission to access the Downloads folder.</string>
         <key>NSRemovableVolumesUsageDescription</key>
         <string>Emacs requires permission to access files on Removable Volumes.</string>
+        <key>NSSpeechRecognitionUsageDescription</key>
+        <string>Emacs requires permission to handle any speech recognition</string>
 </dict>
 </plist>
